go-controller/pkg/node/base_node_network_controller.go

```diff
@@ -15,38 +16,73 @@ import (
 	"k8s.io/klog/v2"
 )
 
-// updateIsOvnUpEnabled checks if ovnkube-master has been upgraded from OvnPortBindingTopoVersion
-// and start to support port binding up field. If so update atomicOvnUpEnabled accordingly.
-func (bnnc *BaseNodeNetworkController) updateIsOvnUpEnabled(ctx context.Context) error {
-	isOvnUpEnabled := atomic.LoadInt32(&bnnc.atomicOvnUpEnabled) > 0
-	if config.OvnKubeNode.Mode != types.NodeModeDPU || isOvnUpEnabled || config.OvnKubeNode.DisableOVNIfaceIdVer {
-		return nil
-	}
+// initUpgradeController creates the upgradeController and return the initial topologyVersion
+func (bnnc *BaseNodeNetworkController) initUpgradeController(ctx context.Context) (*upgrade.UpgradeController, int, error) {
+	// Upgrade for Node. If we upgrade workers before masters, then we need to keep service routing via
+	// mgmt port until masters have been updated and modified OVN config. Run a goroutine to handle this case
 	upgradeController := upgrade.NewController(bnnc.client, bnnc.watchFactory)
 	initialTopoVersion, err := upgradeController.GetTopologyVersion(ctx)
 	if err != nil {
-		return fmt.Errorf("failed to get initial topology version: %w", err)
+		return nil, 0, fmt.Errorf("failed to get initial topology version: %w", err)
 	}
-
 	klog.Infof("Current control-plane topology version is %d", initialTopoVersion)
+	return upgradeController, initialTopoVersion, nil
+}
 
+// updateIsOvnUpEnabledDPUMode is called in DPU mode, it checks if ovnkube-master has been upgraded from
+// OvnPortBindingTopoVersion and start to support port binding up field. If so update atomicOvnUpEnabled accordingly.
+func (bnnc *BaseNodeNetworkController) updateIsOvnUpEnabledDPUMode(ctx context.Context, isOvnUpEnabled bool) error {
```

Comment:
I dont understand why we need `updateIsOvnUpEnabledDPUMode`.

This
```
if config.OvnKubeNode.Mode != types.NodeModeDPU || isOvnUpEnabled || config.OvnKubeNode.DisableOVNIfaceIdVer {
		return nil
	}
```
is already checked in `updateIsOvnUpEnabledCommon`

And this
```
upgradeController, initialTopoVersion, err := bnnc.initUpgradeController(ctx)
```
is certainly used in both in `NodeModeDPU` and `NodeModeFull`

maybe a simple approach would be to have a `handleUpgrade`method that would
```
// return if non dpu or non full mode
// init upgrade controller
// handle isOvnUpEnabled
// return if non full mode
// handle other upgrade stuff
```

Reply:
I have the first check to avoid start the go routine.

Reply:
that is good to know and will simplify the code a lot. I will make the change. Thanks very much!

Reply:
I think it is getting more complicated than needed.

I missunderstood, I though we were talking about the relation between isOvnUpEnabled and upgradeServiceRoute, and not between upgradeServiceRoute and needLegacySvcRoute. 

What would be wrong with what I proposed in https://github.com/ovn-org/ovn-kubernetes/pull/3380#discussion_r1105719829? It is not pretty, but it isn't pretty already in the base code and we would avoid the code duplication.

Or a handle upgrade that would
```
// return if non dpu or non full mode
// init upgrade controller
// hadle needLegacySvcRoute for full mode
// handle isOvnUpEnabled 
// return if non full mode 
// handle upgradeServiceRoute
```

Reply:
@jcaamano updated the code. please take a look. thanks